crates/miden-objects/src/account/storage/map.rs

```diff
@@ -1,37 +1,60 @@
-use miden_crypto::merkle::{EmptySubtreeRoots, MerkleError};
+use alloc::collections::BTreeMap;
+
+use miden_crypto::merkle::EmptySubtreeRoots;
+use vm_core::{EMPTY_WORD, Felt, FieldElement};
 
 use super::{
     ByteReader, ByteWriter, Deserializable, DeserializationError, Digest, Serializable, Word,
 };
 use crate::{
+    Hasher,
     account::StorageMapDelta,
     crypto::{
         hash::rpo::RpoDigest,
         merkle::{InnerNodeInfo, LeafIndex, SMT_DEPTH, Smt, SmtLeaf, SmtProof},
     },
+    errors::StorageMapError,
 };
 
 // ACCOUNT STORAGE MAP
 // ================================================================================================
 
 /// Empty storage map root.
-pub const EMPTY_STORAGE_MAP_ROOT: Digest =
-    *EmptySubtreeRoots::entry(StorageMap::STORAGE_MAP_TREE_DEPTH, 0);
-
-/// Account storage map is a Sparse Merkle Tree of depth 64. It can be used to store more data as
-/// there is in plain usage of the storage slots. The root of the SMT consumes one account storage
-/// slot.
+pub const EMPTY_STORAGE_MAP_ROOT: Digest = *EmptySubtreeRoots::entry(StorageMap::TREE_DEPTH, 0);
+
+/// An account storage map which is a sparse merkle tree of depth [`Self::TREE_DEPTH`] (64).
+///
+/// It can be used to store a large amount of data in an account than would be otherwise possible
+/// using just the account's storage slots. This works by storing the root of the map's underlying
+/// SMT in one account storage slot. Each map entry is a leaf in the tree and its inclusion is
+/// proven while retrieving it (e.g. via `account::get_map_item`).
+///
+/// As a side-effect, this also means that _not all_ entries of the map have to be present at
+/// transaction execution time in order to access or modify the map. It is sufficient if _just_ the
+/// accessed/modified items are present in the advice provider.
+///
+/// Because the keys of the map are user-chosen and thus not necessarily uniformly distributed, the
+/// trees could be imbalanced and made less efficient. To mitigate that, the keys used in the
+/// storage map are hashed before they are inserted into the SMT. The original keys are retained in
+/// a separate map. This causes redundancy but allows for introspection of the map, e.g. by querying
+/// the set of stored (original) keys which is useful in debugging and explorer scenarios.
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct StorageMap {
```

Comment:
Some of this info should probably make it into the mdBook - but maybe this is something we leave for https://github.com/0xPolygonMiden/miden-base/issues/1213.

Reply:
Agreed, I would leave that for that issue.